[ytt] add schema migration guide

gcheadle-vmware opened the pull request.

gcheadle-vmware committed:
[ytt] add schema migration guide

pivotaljohn reviewed (changes requested):
We hit on most the key points. 👍🏻 

But it's landing as particularly dense by my eye... like if I'm in a simpler situation, I can't just read the the first part of the guide and then stop when it gets complex: it's immediately invoking terms like "root library", "private library", 

In order to serve a greater number of users, I think we want to roll it out a bit (thinking pizza dough).

I'm thinking of a progression like this:
1. I've templated my stuff with `ytt` and I have a (i.e. just one) data values file.
2. scenario 1 + now two data files (just like you show it)
3. scenario 1 or 2 + I have dependencies (i.e. other libraries)

and then I can see three how-tos that apply regardless of the scenario:
- How do I mark a section of Data Values as "optional"? (nullable)
- How do I provide default values for an array? (via a Data Values file)
- How do I mark a Data Values as containing any kind of YAML? (any)

If it were broken up to something like that, I think it would be easier for folks from various situations find theirs in the instructions and know when to stop.

pivotaljohn commented on content/ytt/docs/latest/data-values-schema-migration-guide.md:
```diff
@@ -0,0 +1,201 @@
+Now just include the `values-schema.yml` file in your `ytt` invocation instead of the multiple data values files.
+
+### Multiple data values files + Private libraries
+In this scenario, we expect your configuration to depend on a ytt library; outlined in the [library module docs](lang-ref-ytt-library.md).
```
Similar as above: that it's "friendlier" to describe the circumstances from the _reader's_ perspective...

```suggestion
If your configuration depends on a ytt library — outlined in the [library module docs](lang-ref-ytt-library.md) — there are a few points to note.
```

pivotaljohn commented on content/ytt/docs/latest/data-values-schema-migration-guide.md:
````diff
@@ -0,0 +1,201 @@
+  host: registry.dev.io
+  port: 8080
+```
+Now we have migrated the root library to use schemas, and the `ytt` invocation will succeed as the same as before, but the private library, "lib" is not yet migrated to schema. 
````
The fact that `ytt` will work in a mixed-mode is particularly valuable to know. So much so, I'd recommend it gets its own sentence...

```suggestion
Now we have migrated the root library to use schemas, and the `ytt` invocation will succeed as before. Each library can independently opt-in to using schemas.

In this example, we will migrate all libraries. Here, the private library, "lib" is not yet migrated to schema.
```

pivotaljohn commented on content/ytt/docs/latest/data-values-schema-migration-guide.md:
```diff
@@ -0,0 +1,201 @@
+Now simply follow the steps in either of the previous examples to migrate the private library to use schemas.
+
+## How do I provide default values for an array?
+Arrays in schemas have [special behavior](lang-ref-ytt-schema.md#inferring-defaults-for-arrays), their default value is empty, but their inferred type is from the value in the schema document. 
+Arrays must have one and only one item, which provides the inferred type of that data value. The example below shows how to define an array in a schema and them provide default values via a data values file.
+
```
This is a tricky set of facts to unfurl. 😅 

In a somewhat complex situation — like this one — getting nit-picky precise about what everything is can help avoid confusion.

- _pedantically_, it's the type of the _elements_ that's being declared.
- "inferred" is slightly more accurate as a verb than an adjective, here.
- lead in with the fact that only one value is specified for arrays — this makes it possible to "chronologically" connect this set of related facts.

```suggestion
Arrays in schemas have [special behavior](lang-ref-ytt-schema.md#inferring-defaults-for-arrays): exactly one element is specified in the schema of an array, and that value is _only_ used to infer the type of that array's elements — the default value for arrays is always empty.

The example below shows how to define an array in a schema and then provide default values via a data values file.
```

👍🏻  The link to the section in schema reference page is _**critical**_ here. Great pointer.

pivotaljohn commented on content/ytt/docs/latest/data-values-schema-migration-guide.md:
```diff
@@ -0,0 +1,201 @@
+
+
+## How do I mark a section of Data Values as "optional"?
+The [@schema/nullalble](lang-ref-ytt-schema.md#schemanullable) annotation can be used to override the defaults of the node it annotates.
+
+Use this annotation on a node if `null` is valid input, or you want the default value for that data value to be `null`. (This annotation cannot be use on arrays.)
```
A combination of suggestions:

- lead with describing the user's experience.
- in order to communicate how the concept of "optional" is realized, I think we need to show _both_ how the section is annotated _and_ how it is used.
- seems like an example would be the most terse/effective way to show that.
- that array elements can't be annotated as null is better described in reference material.

```suggestion
Sometimes your configuration includes a section of Data Values that are not typically used or in some way optional.

The [@schema/nullable](lang-ref-ytt-schema.md#schemanullable) annotation can be used to mark such a section as `null`.

        ```yaml
        #@schema/nullable
        aws:
          username: admin
          password: "1234"
        ```

The default value for `data.values.aws` will be `null`.

`null` is falsey.  Templates using that optional data value can perform a simple check for whether the section was specified:

        ```yaml
        #@ load("@ytt:data", "data")
        ---
        #@ if/end data.values.aws:
        s3_creds:
          user: #@ data.values.aws.username
          password: #@ data.values.aws.password
        ```
```

Note: the code examples do _not_ have to be indented... I had to do that in order [to get the full "suggestion" to display in this here GitHub comment](https://github.com/vmg/redcarpet/issues/208#issuecomment-14738077).

pivotaljohn reviewed (changes requested):
Solid iteration. This is much more approachable. 👍🏻 

I think we're essentially there, but I think a few changes are necessary.

Check out the details, below.

Very happy to chat about this stuff too.

gcheadle-vmware committed:
Address feedback

- multiple data values documents = multiple schema documents
- include readers perspective

pivotaljohn reviewed (approved):
Once again, great work, @gcheadle-vmware.

It's an unsung challenge to mix in the right ingredients in the right amounts to yield just the right amount of detail, written from the right perspective, in the right voice.

👍🏻

pivotaljohn commented:
I forgot to delay merging doc changes. This was supposed to be merged with other schema documentation changes.

However, it's (almost) purely additive, so it's okay in this case.
